rpcs/ledgerService.go

```diff
@@ -161,189 +134,37 @@ func (ls *LedgerService) ServeHTTP(response http.ResponseWriter, request *http.R
 			response.WriteHeader(http.StatusInternalServerError)
 			return
 		}
 	}
-
+	defer cs.Close()
 	response.Header().Set("Content-Type", LedgerResponseContentType)
-	response.Header().Set("Content-Length", strconv.Itoa(len(encodedBlockCert)))
-	response.Header().Set("Cache-Control", ledgerResponseHasBlockCacheControl)
-	response.WriteHeader(http.StatusOK)
-	_, err = response.Write(encodedBlockCert)
-	if err != nil {
-		logging.Base().Warn("http block write failed ", err)
-	}
-}
-
-// WsGetBlockRequest is a msgpack message requesting a block
-type WsGetBlockRequest struct {
-	Round uint64 `json:"round"`
-}
-
-// WsGetBlockOut is a msgpack message delivered on responding to a block (not rpc-based though)
-type WsGetBlockOut struct {
-	Round      uint64
-	Error      string
-	BlockBytes []byte `json:"blockbytes"`
-}
-
-func (ls *LedgerService) processIncomingMessage(msg network.IncomingMessage) (n network.OutgoingMessage) {
-	// don't block - just stick in a slightly buffered channel if possible
-	select {
-	case ls.catchupReqs <- msg:
-	default:
-	}
-	// don't return outgoing message, we just unicast instead
-	return
-}
-
-// ListenForCatchupReq handles catchup getblock request
-func (ls *LedgerService) ListenForCatchupReq(reqs <-chan network.IncomingMessage, stop chan struct{}) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	for {
-		select {
-		case reqMsg := <-reqs:
-			ls.handleCatchupReq(ctx, reqMsg)
-		case <-stop:
-			return
-		}
-	}
-}
-
-const noRoundNumberErrMsg = "can't find the round number"
-const noDataTypeErrMsg = "can't find the data-type"
-const roundNumberParseErrMsg = "unable to parse round number"
-const blockNotAvailabeErrMsg = "requested block is not available"
-const datatypeUnsupportedErrMsg = "requested data type is unsupported"
-
-// a blocking function for handling a catchup request
-func (ls *LedgerService) handleCatchupReq(ctx context.Context, reqMsg network.IncomingMessage) {
-	var res WsGetBlockOut
-	target := reqMsg.Sender.(network.UnicastPeer)
-	var respTopics network.Topics
-
-	if target.Version() == "1" {
-
-		defer func() {
-			ls.sendCatchupRes(ctx, target, reqMsg.Tag, res)
-		}()
-		var req WsGetBlockRequest
-		err := protocol.DecodeReflect(reqMsg.Data, &req)
-		if err != nil {
-			res.Error = err.Error()
-			return
-		}
-		res.Round = req.Round
-		encodedBlob, err := RawBlockBytes(ls.ledger, basics.Round(req.Round))
-
-		if err != nil {
-			res.Error = err.Error()
-			return
-		}
-		res.BlockBytes = encodedBlob
+	requestedCompressedResponse := strings.Contains(request.Header.Get("Accept-Encoding"), "gzip")
+	if requestedCompressedResponse {
+		response.Header().Set("Content-Encoding", "gzip")
+		io.Copy(response, cs)
 		return
 	}
-	// Else, if version == 2.1
-	defer func() {
-		target.Respond(ctx, reqMsg, respTopics)
-	}()
-
-	topics, err := network.UnmarshallTopics(reqMsg.Data)
+	decompressedGzip, err := gzip.NewReader(cs)
```

Comment:
Is this necessary? Why not restrict only to gzip streams?
I worry that this may be very expensive, and overwhelm the server even within the DOS prevention filter limits.

Reply:
It is necessary to support "classic" http streams so that we could tunnel the stream through proxies and other storage providers in a http compliant way. Not all the storage providers and/or proxy/firewall providers supports compressed streams. By supporting both on either ends we're covering all the cases.
I don't think that the DoS risk here is that great. The load on the server is incremental - i.e. it will perform the decompression only as long as the other side is receiving the bytes.
Also, the compression rate here is not malicious ( since we've created this file), so that won't be an issue.